contracts/leaser/src/migrate.rs

```diff
@@ -2,66 +2,97 @@ use lease::api::MigrateMsg;
-    let migrated_msgs = leases.fold(no_msgs, MigrateBatch::migrate_lease);
+    let no_msgs = MigrateBatch::new(lease_code_id, max_leases);
+    let migrated_msgs = leases
+        .take(max_leases.try_into()?)
+        .fold(no_msgs, MigrateBatch::migrate_lease);
     migrated_msgs.try_into()
 }
 
+impl MigrationResult {
+    pub fn try_add_msgs<F>(mut self, add_fn: F) -> ContractResult<Self>
+    where
+        F: FnOnce(&mut Batch) -> ContractResult<()>,
+    {
+        add_fn(&mut self.msgs).map(|()| self)
+    }
+}
 struct MigrateBatch {
     new_code_id: u64,
-    may_batch: ContractResult<Batch>,
+    leases_left: NbInstances,
+    may_result: ContractResult<MigrationResult>,
 }
 impl MigrateBatch {
-    fn new(new_code_id: u64) -> Self {
+    fn new(new_code_id: u64, max_leases: NbInstances) -> Self {
         Self {
             new_code_id,
-            may_batch: Ok(Batch::default()),
+            leases_left: max_leases,
+            may_result: Ok(Default::default()),
         }
     }
 
     fn migrate_lease(mut self, lease_contract: ContractResult<Addr>) -> Self {
-        let op = |mut batch: Batch| {
+        let op = |result: MigrationResult| {
             lease_contract.and_then(|lease| {
-                batch
-                    .schedule_migrate_wasm_no_reply(&lease, MigrateMsg {}, self.new_code_id)
-                    .map(|_| batch)
-                    .map_err(Into::into)
+                self.leases_left -= 1;
+                result
+                    .try_add_msgs(|msgs| {
+                        msgs.schedule_migrate_wasm_no_reply(&lease, MigrateMsg {}, self.new_code_id)
+                            .map_err(Into::into)
+                    })
+                    .map(|mut res| {
+                        if self.leases_left == NbInstances::MIN {
+                            res.last_instance = Some(lease);
+                        }
+                        res
+                    })
             })
         };
 
-        self.may_batch = self.may_batch.and_then(op);
+        self.may_result = self.may_result.and_then(op);
         self
     }
 }
 
-impl TryFrom<MigrateBatch> for Batch {
+impl TryFrom<MigrateBatch> for MigrationResult {
     type Error = ContractError;
     fn try_from(this: MigrateBatch) -> Result<Self, Self::Error> {
-        this.may_batch
+        this.may_result
     }
 }
 
 #[cfg(test)]
 mod test {
     use lease::api::MigrateMsg;
-    use platform::batch::Batch;
     use sdk::cosmwasm_std::Addr;
 
-    use crate::ContractError;
+    use crate::{migrate::MigrationResult, ContractError};
 
     #[test]
     fn no_leases() {
         let new_code = 242;
         let no_leases = vec![];
         assert_eq!(
-            Ok(Batch::default()),
-            super::migrate_leases(no_leases.into_iter().map(Ok), new_code)
+            Ok(MigrationResult::default()),
+            super::migrate_leases(no_leases.into_iter().map(Ok), 2, new_code)
         );
```

Comment:
Expected expressions are second arguments to the macro.
```suggestion
        assert_eq!(
            super::migrate_leases(no_leases.into_iter().map(Ok), 2, new_code),
            Ok(MigrationResult::default())
        );
```

Reply:
same